### PR TITLE
Support multi-stream communication for dynamic graph distributed

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Performance optimization
### PR changes

Others
### Describe

Support multi-stream communication for dynamic graph distributed.
image

image

image

### CODE REVIEW

File: paddle/fluid/imperative/nccl_context.cc [hunk 1]
```diff
@@ -85,27 +87,37 @@ void NCCLParallelContext::RecvNCCLID(const std::string &ep,
...
   std::string host = addr[0];
   int port = std::stoi(addr[1]);
-  // struct sockaddr_in address;
   int sock = 0;
   struct sockaddr_in serv_addr;
   char buffer[1024] = {0};
 
-  memcpy(buffer, nccl_id, NCCL_UNIQUE_ID_BYTES);
+  PADDLE_ENFORCE_LE(
+      nrings, 8, platform::errors::PreconditionNotMet(
```
Comment (hutuxian) [inline paddle/fluid/imperative/nccl_context.cc]: What about making this error message earlier? Such as in the python code where getting the environment variable.

The nrings of nccl should be less than or equal to 8, which is enough in most scenarios. And your nrings is %d"

File: paddle/fluid/imperative/nccl_context.cc [hunk 2]
```diff
@@ -85,27 +87,37 @@ void NCCLParallelContext::RecvNCCLID(const std::string &ep,
...
 
   VLOG(3) << "closing the socket server: " << ep;
   close(server_fd);
 }
 
-void NCCLParallelContext::SendNCCLID(const std::string &ep,
-                                     ncclUniqueId *nccl_id) {
+void NCCLParallelContext::SendNCCLID(
+    const std::string &ep,
+    std::vector<ncclUniqueId> &nccl_ids) {  // NOLINT
```
Comment (hutuxian) [inline paddle/fluid/imperative/nccl_context.cc]: const?

File: paddle/fluid/imperative/reducer.cc [hunk 3]
```diff
@@ -310,29 +315,34 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
   for (; next_group_ < groups_.size() && groups_[next_group_].pending_ == 0;
        ++next_group_) {
     auto &group = groups_[next_group_];
+    int run_order = next_group_ % nrings_;
     if (group.is_sparse_) {
-      VLOG(3) << "sparse group [" << next_group_ << "] start allreduce...";
-      parallel_ctx_->AllReduceByStream(*group.sparse_contents_,
-                                       group.sparse_contents_, 0, false);
+      VLOG(3) << "sparse group [" << next_group_
+              << "] start allreduce in order[" << run_order << "]";
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: in order[x] -> in ring[x]? Same as line 333

File: paddle/fluid/imperative/reducer.cc [hunk 4]
```diff
@@ -310,29 +315,34 @@ void Reducer::MarkGroupReady(size_t group_index) {
   }
 
   PADDLE_ENFORCE_CUDA_SUCCESS(
-      cudaEventRecord(events_[group_index].get(), compute_stream_));
-  PADDLE_ENFORCE_CUDA_SUCCESS(
-      cudaStreamWaitEvent(comm_stream_, events_[group_index].get(), 0));
+      cudaEventRecord(group_events_[group_index].get(), compute_stream_));
+  for (int i = 0; i < nrings_; ++i)
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: add `{}` after `for`?

File: paddle/fluid/imperative/nccl_context.cc (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/imperative/nccl_context.cc]: ok. fix it .

File: paddle/fluid/imperative/nccl_context.cc (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/imperative/nccl_context.cc]: ok, fix it

File: paddle/fluid/imperative/reducer.cc (hunk 3 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc]: ok, i will fix it.

File: paddle/fluid/imperative/reducer.cc (hunk 4 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc]: ok, i will fix it.

**Review by hutuxian — APPROVED** [review]
LGTM

**Pull request merged**